### PR TITLE
Launch an optimization with a relative margin objective function

### PR DESCRIPTION
**Pull request by pet-mit**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
New feature: optimizing the relative margin

**What is the current behavior?** *(You can also link to an open issue here)*
Only absolute margins are maximized

**What is the new behavior (if this is a feature change)?**
The user can choose to maximize the minimum relative margin when all margins are positive

### CODE REVIEW

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoUtil.java [hunk 1]
```diff
@@ -83,6 +99,8 @@ public static SystematicSensitivityInterface createSystematicSensitivityInterfac
 
             // We may also want to have a different interface for the first run and the successive runs if we do not wish to
             // compute the PTDFs at every iteration.
+        } else if (raoParameters.getObjectiveFunction().doesRequirePtdf()) {
+            builder.withPtdfSensitivities(raoData.getGlskProvider(), raoData.getCnecs());
```
Comment (phiedw) [inline ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoUtil.java:103]: this shouldn't be required if we only wish to compute the PTDFs at the start of the RAO

**Review by phiedw — CHANGES_REQUESTED** [review]

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/AbsolutePtdfSumsComputation.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                if (ptdfMap.get(cnec.getId()).containsKey(countryPair.getLeft()) && ptdfMap.get(cnec.getId()).containsKey(countryPair.getRight())) {
+                    ptdfSum += Math.abs(ptdfMap.get(cnec.getId()).get(countryPair.getLeft()).doubleValue() - ptdfMap.get(cnec.getId()).get(countryPair.getRight()).doubleValue());
+                }
+            }
+            ptdfSums.put(cnec.getId(), ptdfSum);
+        });
+        return ptdfSums;
+    }
+
+    private static Map<String, Map<Country, Double>> computePtdf(Set<Cnec> cnecs, Network network, GlskProvider glskProvider, SystematicSensitivityResult ptdfsAndRefFlows) {
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/AbsolutePtdfSumsComputation.java]: done

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/objective_function_evaluator/MinMarginObjectiveFunction.java [hunk 3]
```diff
@@ -22,38 +22,58 @@
...
     }
 
+    public boolean isRelative() {
+        return relativeMargin;
+    }
+
     @Override
     public double getFunctionalCost(RaoData raoData) {
-        return minMarginEvaluator.getCost(raoData);
+        double inverseMinimumMargin = minMarginEvaluator.getCost(raoData);
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/objective_function_evaluator/MinMarginObjectiveFunction.java]: yes !

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoUtil.java (hunk 1 above)
Comment (pet-mit) [reply ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoUtil.java]: removed, it was unnecessary indeed

**Review by phiedw — APPROVED** [review]

**Pull request merged**